x/ibc/light-clients/07-tendermint/types/proposal_handle.go

```diff
@@ -12,116 +10,114 @@ import (
-	if err := cs.checkProposedHeader(consensusState, header, currentTimestamp); err != nil {
-		return nil, nil, err
 	}
 
-	newClientState, consensusState := update(ctx, clientStore, &cs, header)
-	return newClientState, consensusState, nil
-}
-
-// checkProposedHeader checks if the Tendermint header is valid for updating a client after
-// a passed proposal.
-// It returns an error if:
-// - the header provided is not parseable to tendermint types
-// - header height is less than or equal to the latest client state height
-// - signed tendermint header is invalid
-// - header timestamp is less than or equal to the latest consensus state timestamp
-// - header timestamp is expired
-// NOTE: header.ValidateBasic is called in the 02-client proposal handler. Additional checks
-// on the validator set and the validator set hash are done in header.ValidateBasic.
-func (cs ClientState) checkProposedHeader(consensusState *ConsensusState, header *Header, currentTimestamp time.Time) error {
-	tmSignedHeader, err := tmtypes.SignedHeaderFromProto(header.SignedHeader)
-	if err != nil {
-		return sdkerrors.Wrap(err, "signed header in not tendermint signed header type")
-	}
+	cs.LatestHeight = substituteClientState.LatestHeight
 
-	if !header.GetTime().After(consensusState.Timestamp) {
-		return sdkerrors.Wrapf(
-			clienttypes.ErrInvalidHeader,
-			"header timestamp is less than or equal to latest consensus state timestamp (%s ≤ %s)", header.GetTime(), consensusState.Timestamp)
+	// validate the updated client and ensure it isn't expired
+	if err := cs.Validate(); err != nil {
+		return nil, sdkerrors.Wrap(err, "unexpected error: updated subject client state is invalid")
 	}
 
-	// assert header height is newer than latest client state
-	if header.GetHeight().LTE(cs.GetLatestHeight()) {
-		return sdkerrors.Wrapf(
-			clienttypes.ErrInvalidHeader,
-			"header height ≤ consensus state height (%s ≤ %s)", header.GetHeight(), cs.GetLatestHeight(),
+	latestConsensusState, err := GetConsensusState(subjectClientStore, cdc, cs.GetLatestHeight())
+	if err != nil {
+		return nil, sdkerrors.Wrapf(
+			err, "unexpected error: could not get consensus state for updated subject client from clientstore at height: %d", cs.GetLatestHeight(),
 		)
 	}
 
-	if err := tmSignedHeader.ValidateBasic(cs.GetChainID()); err != nil {
-		return sdkerrors.Wrap(err, "signed header failed basic validation")
+	if cs.IsExpired(latestConsensusState.Timestamp, ctx.BlockTime()) {
+		return nil, sdkerrors.Wrap(clienttypes.ErrInvalidClient, "updated subject client is expired")
 	}
 
-	if cs.IsExpired(header.GetTime(), currentTimestamp) {
-		return sdkerrors.Wrap(clienttypes.ErrInvalidHeader, "header timestamp is already expired")
-	}
+	return &cs, nil
+}
 
-	return nil
+// IsMatchingClientState returns true if all the client state parameters match
+// except for frozen height, latest height, and chain-id.
+func IsMatchingClientState(subject, substitute ClientState) bool {
```

Comment:
This function should also ignore all custom fields in order to support updating multiple subject clients with the same substitute.

Reply:
what do you define as custom fields? Just the Allow booleans? I think `TrustingPeriod` needs to be identical since that changes the possible consensus states allowed to be committed for that client state. While technically the security model allows us to do anything, it would be strange to me if a client state contains consensus states that violate its trusting period

Reply:
I personally think this is fine. Yes this could theoretically mean that the consensus states get filled in such that the Trusting period is not respected. But since we're filling in past consensus states and governance is trusting the substitute client, I'm not too worried about it.

If you want to be really fancy, I guess you could make sure that you don't fill in consensus states that are "unexpired" and thus give the clients the chance to fill in current consensus states according to their own trusting period. But i think that's unnecessarily complex.

We will do this anyway if we support updating to a different revision, as the first consensus state of the next revision may not be within the trusting period of the last consensus state on the subject client.

cc: @cwgoes

Reply:
I do not think the `Allow` booleans should be allowed to change through the mechanism for updating expired or invalidated clients, since those booleans are chosen by the client creator according to their preferences when creating the client, there's no reason to change them when an irregular update happens - a separate client should be created if someone wants to use different parameters of that kind.

I don't think it is especially dangerous to allow the trusting period to be changed in this case, since manual governance can switch the client to a totally different header, chain ID, etc. anyways, but it still seems a little strange to me _for_ it to be changed in this mechanism - it strikes me as in the same "category" as the allow booleans - relating to original intent of the client creator, where different clients for the same chain with different trusting periods could be created if users have different preferences.

I think it may be worth distinguishing between "governance override due to misbehaviour" - trusting period should _not_ change - and "governance override" in general, if the client selects that - trusting period should be allowed to change - here we're talking just about unfreezing due to misbehaviour, right?

Thoughts? Do you agree with that distinction? Are there particular advantages to allowing the trusting period to change that I've missed?

Reply:
> I do not think the Allow booleans should be allowed to change through the mechanism for updating expired or invalidated clients, since those booleans are chosen by the client creator according to their preferences when creating the client, there's no reason to change them when an irregular update happens - a separate client should be created if someone wants to use different parameters of that kind.

None of these custom fields (including TrustingPeriod) would change through these mechanisms. These original fields would be copied over from the original client.

The question is whether the substitute client **must** share the same custom fields as the subject client.

If we don't enforce this, then multiple expired clients (of the same counterparty chain) can be unfrozen by the same substitute client even though they have different custom parameters. Note that the custom parameters of each individual client will be persisted across the governance override.

The only problem with this solution is if the substitute client has a higher trusting period than a subject client.
In this case, when the consensus states are copied over from the substitute client to the subject client, the consensus states may be spread out in such a way that they would have been considered __invalid__ by the subject client through the normal mechanism since neighboring consensus states may be further apart than the subject client's TrustingPeriod (but within the substitute client's larger TrustingPeriod).

Note this doesn't have any implications for correctness or security, but it does break the invariant that neighboring consensus states will be at most a `TrustingPeriod` apart from each other in terms of timestamp.

This can be broken since they may be filled in by a more lenient substitute client during the governance override.

I don't believe this is an issue personally since governance is approving this more "lax" approach to filling in past consensus states anyway. We could make it such that we don't fill in consensus states that could be filled in by a relayer which would enforce the TrustingPeriod (thus preserving the invariant for current consensus states)

Also, this invariant could be broken anyway if the governance override happens across an upgrade and the `InitialHeight` consensus state of the new revision is past the `TrustingPeriod` of the last consensus state stored on the subject client.

Reply:
Sorry, I misunderstood how the substitute client process works. Thanks for the clarification.

> it does break the invariant that neighboring consensus states will be at most a TrustingPeriod apart from each other in terms of timestamp.

I can't think of any immediate problems that this would cause, but it still is a little odd. Let's at least note this somewhere.

Reply:
I'm going to open a separate issue for multiple client functionality. It'll be easier to review its correctness and it isn't a necessary feature (so no need to delay merging this pr). I also would like to test a governance proposal with a single client to find out how much [gas is used](https://github.com/cosmos/cosmos-sdk/pull/8405#discussion_r571124823) (assuming someone is trying to block the proposal by updating the substitute with the maximum possible consensus states)

Reply:
made #8558